src/verify.rs

```diff
@@ -101,18 +101,26 @@ pub struct StackSlot {
 pub trait ProofBuilder {
     /// The data type being generated
     type Item: Clone;
+    /// The hyp gathering type
+    type Accum: Default;
+
+    /// Add a new hyp to the accumulation type
+    fn push(&mut self, hyps: &mut Self::Accum, hyp: Self::Item);
 
     /// Create a proof data node from a statement, the data for the hypotheses, and the compressed constant string
-    fn build(&mut self, addr: StatementAddress, hyps: Vec<Self::Item>, expr: &[u8]) -> Self::Item;
+    fn build(&mut self, addr: StatementAddress, hyps: Self::Accum, expr: &[u8]) -> Self::Item;
 }
 
 /// The "null" proof builder, which creates no extra data. This
 /// is used for one-shot verification, where no extra data beyond the stack
 /// information is needed.
 impl ProofBuilder for () {
     type Item = ();
+    type Accum = ();
+
+    fn push(&mut self, _: &mut (), _: ()) {}
```

Comment:
The main case where you positively need `#[inline]` is cross-crate (it triggers the compiler to save bitcode, otherwise cross-crate inlining is impossible).  I'm pretty sure an empty method will always be inlined same-crate.

It's unclear to me whether plain `#[inline]` has an effect within one crate.